packages/wallets/src/features/cashier/modules/Transfer/components/TransferFormInputField/TransferFormInputField.tsx

```diff
@@ -1,29 +1,101 @@
+import { WalletText } from '../../../../../../components/Base';
 import useInputATMFormatter from '../../../../../../hooks/useInputATMFormatter';
 import './TransferFormInputField.scss';
 
 type TProps = {
-    defaultValue?: number;
-    fieldName: string;
+    currency?: string;
+    disabled?: boolean;
     fractionDigits?: number;
     label: string;
+    locale?: Intl.LocalesArgument;
+    maxDigits?: number;
+    onChange?: (value: number) => void;
+    value: number;
 };
 
-const WalletTransferFormInputField: React.FC<TProps> = ({ defaultValue, fieldName, fractionDigits = 2, label }) => {
-    const { setFieldValue } = useFormikContext();
-    const { onChange, value } = useInputATMFormatter(defaultValue, {
+const WalletTransferFormInputField: React.FC<TProps> = ({
+    currency,
+    disabled,
+    fractionDigits = 0,
+    label,
+    locale,
+    maxDigits,
+    onChange,
+    value,
+}) => {
+    const inputRef = useRef<HTMLInputElement>(null);
+    const input = inputRef.current;
+
+    const [isFocused, setIsFocused] = useState<boolean>(false);
+    const [caretNeedsRepositioning, setCaretNeedsRepositioning] = useState<boolean>(false);
+    const [caret, setCaret] = useState<number>();
+
+    const {
+        onChange: formatOnChange,
+        onPaste: formatOnPaste,
+        value: formattedValue,
+    } = useInputATMFormatter(value, {
         fractionDigits,
+        locale,
     });
+    const [prevFormattedValue, setPrevFormattedValue] = useState<string>(formattedValue);
 
     useEffect(() => {
-        setFieldValue(fieldName, Number(value));
-    }, [fieldName, setFieldValue, value]);
+        setPrevFormattedValue(formattedValue);
+        onChange?.(Number(unFormatLocaleString(formattedValue, locale)));
+    }, [caretNeedsRepositioning, formattedValue, locale, onChange]);
+
+    // keep the caret from jumping
+    useEffect(() => {
+        if (caret && caretNeedsRepositioning) {
+            // if next to a comma or period, prefer positioning the caret left to it
+            const newCaretPosition = /[,.]/g.test(formattedValue[formattedValue.length - 1 - caret])
+                ? formattedValue.length - caret - 1
+                : formattedValue.length - caret;
+            input?.setSelectionRange(newCaretPosition, newCaretPosition);
+            setCaretNeedsRepositioning(false);
+        }
+    }, [caret, formattedValue, caretNeedsRepositioning, input]);
+
+    // override some editing behavior for better UX
+    const onChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+        if (!input) return;
+        setCaret(input.value.length - (input.selectionStart ?? 0));
+        setCaretNeedsRepositioning(true);
+        if (maxDigits && input.value.replace(/[.,]/g, '').length > maxDigits) return;
+        if (
+            input.value.length + 1 === prevFormattedValue.length &&
+            input.value.replaceAll(/[,.]/g, '') === prevFormattedValue.replaceAll(/[,.]/g, '')
+        )
```

Comment:
Isnt it better to handle this in the hook `useInputATMFormatter`

Reply:
To be honest, I'm not sure, this would require passing the input's ref into the hook + might make the hook to logic-heavy 🤔